pyadjoint/optimization/tao_solver.py

```diff
@@ -0,0 +1,491 @@
+                raise RuntimeError("Unable to instantiate new dual space "
+                                   "object")
+        return tuple(m_dual._ad_copy() for m_dual in M_dual)
+
+
+class TAOSolver(OptimizationSolver):
+    """Use TAO to solve an optimization problem.
+
+    Args:
+        problem (MinimizationProblem): Defines the optimization problem to be
+            solved.
+        parameters (Mapping): TAO options.
+        comm (petsc4py.PETSc.Comm or mpi4py.MPI.Comm): Communicator.
+        convert_options (Mapping): Defines the `options` argument to
+            :meth:`OverloadedType._ad_convert_type`.
+    """
+
+    def __init__(self, problem, parameters, *, comm=None, convert_options=None):
+        if PETSc is None:
+            raise RuntimeError("PETSc not available")
+
+        if not isinstance(problem, MinimizationProblem):
+            raise TypeError("MinimizationProblem required")
+        if problem.constraints is not None:
+            raise NotImplementedError("Constraints not implemented")
+
+        if comm is None:
+            comm = PETSc.COMM_WORLD
+        if hasattr(comm, "tompi4py"):
+            comm = comm.tompi4py()
+        if convert_options is None:
+            convert_options = {}
+
+        taoobjective = TAOObjective(problem.reduced_functional)
+
+        vec_interface = PETScVecInterface(
+            tuple(m.control for m in taoobjective.reduced_functional.controls),
+            comm=comm)
+        n, N = vec_interface.n, vec_interface.N
+        to_petsc, from_petsc = vec_interface.to_petsc, vec_interface.from_petsc
+        tao = PETSc.TAO().create(comm=comm)
+
+        def objective_gradient(tao, x, g):
+            M = taoobjective.new_M()
+            from_petsc(x, M)
+            J_val, dJ = taoobjective.objective_gradient(M)
+            to_petsc(g, dJ)
+            return J_val
+
+        tao.setObjectiveGradient(objective_gradient, None)
+
+        def hessian(tao, x, H, P):
+            H.getPythonContext().set_M(x)
+
+        class Hessian:
+            def __init__(self):
+                self._shift = 0.0
+
+            @cached_property
+            def _M(self):
+                return taoobjective.new_M()
+
+            def set_M(self, x):
+                from_petsc(x, self._M)
+                self._shift = 0.0
+
+            def shift(self, A, alpha):
+                self._shift += alpha
+
+            def mult(self, A, x, y):
+                M_dot = taoobjective.new_M()
+                from_petsc(x, M_dot)
+                ddJ = taoobjective.hessian(self._M, M_dot)
+                to_petsc(y, ddJ)
+                if self._shift != 0.0:
+                    y.axpy(self._shift, x)
+
+        H_matrix = PETSc.Mat().createPython(((n, N), (n, N)),
+                                            Hessian(), comm=comm)
+        H_matrix.setOption(PETSc.Mat.Option.SYMMETRIC, True)
+        H_matrix.setUp()
+        tao.setHessian(hessian, H_matrix)
+
+        class GradientNorm:
```

Comment:
also docstring saying what this thing implements.

Reply:
Done.